client/consensus/grandpa/authorities.go

```diff
@@ -160,51 +231,414 @@ func (authSet *AuthoritySet) addForcedChange(pending PendingChange, isDescendent
+
+		// TODO add telemetry
+
+		// Store the set_id together with the last block_number for the set
+		authoritySetChange := authorityChange{
+			setId:       authSet.setId,
+			blockNumber: finalizedNumber,
+		}
+		authSet.authoritySetChanges = append(authSet.authoritySetChanges, authoritySetChange)
+		authSet.currentAuthorities = finalizationResult.value.nextAuthorities
+		authSet.setId++
+
+		status.newSetBlock = &newSetBlockInfo{
+			newSetBlockNumber: finalizedNumber,
+			newSetBlockHash:   finalizedHash,
+		}
+	}
+
+	return status, nil
+}
+
+// EnactsStandardChange Check whether the given finalized block number enacts any standard
+// authority set change (without triggering it), ensuring that if there are
+// multiple changes in the same branch, finalizing this block won't
+// finalize past multiple transitions (i.e. transitions must be finalized
+// in-order). Returns *true if the block being finalized enacts a
+// change that can be immediately applied, *false if the block being
+// finalized enacts a change but it cannot be applied yet since there are
+// other dependent changes, and nil if no change is enacted. The given
+// function `is_descendent_of` should return `true` if the second hash
+// (target) is a descendent of the first hash (base).
+func (authSet *AuthoritySet) EnactsStandardChange(
+	finalizedHash common.Hash, finalizedNumber uint, isDescendentOf IsDescendentOf) (*bool, error) {
+	applied, err := authSet.pendingStandardChanges.FinalizeAnyWithDescendentIf(&finalizedHash, finalizedNumber, isDescendentOf, enactStandardChangesPredicate(finalizedNumber))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errForkTree, err)
+	}
+	return applied, nil
 }
 
-// InnerLocked
-// Returns access to the [`AuthoritySet`] and locks it.
+// PendingChange A pending change to the authority set.
 //
-// For more information see [`SharedDataLocked`].
-// TODO update with shared logic
-func (sas *SharedAuthoritySet) InnerLocked() AuthoritySet {
-	return sas.authoritySet
+// This will be applied when the announcing block is at some depth within
+// the finalized or unfinalized chain.
+type PendingChange struct {
+	// The new authorities and weights to apply.
+	nextAuthorities AuthorityList
+	// How deep in the chain the announcing block must be
+	// before the change is applied.
+	delay uint
+	// The announcing block's height.
+	canonHeight uint
+	// The announcing block's hash.
+	canonHash common.Hash
+	// The delay kind.
+	delayKind DelayKind
+}
+
+// EffectiveNumber Returns the effective number this change will be applied at.
+func (pc *PendingChange) EffectiveNumber() uint {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced change.
+type AuthoritySetChanges []authorityChange
```

Comment:
Do we clear/prune these auth changes stored at some point? As I see we append and sometimes we iterate over them from the very first item if we don't clear/prune the complexity will grow linearly and in long operations the node can slow down waiting for the iteration to finish, right?

Reply:
This is a good point. I think potentially beyond scope of this PR, since I was referencing substrate (unless I missed this functionality), but I think this is a good point and something we should think about. Will tag @timwu20 here to get his thoughts on this when he reviews, but my take is that we should do this but outside of this PR (and probably once we build out more critical grandpa components)

Reply:
is this outdated now?